### ISSUE TITLE
Update to serde 0.9

### ISSUE THREAD

**Issue by sfackler** [alexcrichton/toml-rs#135]

**Comment by sfackler** [comment]
There are a couple of behavior changes that this upgrade will mandate:

1. The current enum deserialization approach of trying each variant in turn is no longer possible, since visitors are passed by value. I believe this functionality is only really used to allow a value to be deserialized from multiple types at once, which is possible in serde via a custom `Deserialize` implementation.
2. The current implementation saves off unknown fields when deserializing, which is again no longer possible since visitors are passed by value. I believe the old functionality can be retrieved via a `DeserializeSeed` implementation?

**Comment by dtolnay** [comment]
The enum part I intend to address in serde_derive within the next week. As Steven said, it's just a matter of generating slightly different Deserialize code to try all the cases. This is not something that needs to be built into every Deserializer individually.

Let me know if you get stuck on the unknown fields update and we can figure something out.

**Comment by alexcrichton** [comment]
Yeah so to clarify the two desires of the `rustc-serialize` implementation (which spilled over to the serde impl) is to support two features:

1. Values which can be configured via a number of means. For example in Cargo you can specify a dependency as `foo = "0.3"` but you can also specify a dependency as `foo = { version = "0.3" }` (with other info). To do this Cargo [has an enum](https://github.com/rust-lang/cargo/blob/8d5383866e387a5bf88e25181d23c1af67cb3abe/src/cargo/util/toml.rs#L212-L216) representing the two forms of configuration, and those are tried in turn to take the first successful one.
2. Cargo supports emitting warnings on unused fields. This basically is a way of slurping up a TOML file and then being able to print out "these fields are unused" to catch typos and such.

Bumping the serde version will require a major version bump of this crate anyway, so I'm all for piling on more breakage as well! In that sense I'm totally cool with whatever's best to support these desires. @dtolnay are you thinking like a custom serde attribute on the [enum I mentioned](https://github.com/rust-lang/cargo/blob/8d5383866e387a5bf88e25181d23c1af67cb3abe/src/cargo/util/toml.rs#L212-L216) to indicate the particular deserialization behavior of "first successful variant wins"? If so, that sounds great!

Tracking used/unused fields is kinda wonky today anyway, but I suspect the ownership-threading changes in 0.9 won't impact that, so I'm sure we can still finagle.

All of this sort makes me think that I should just release toml 0.3 and drop rustc-serialize support at the same time. Serde is the future!

**Comment by sfackler** [comment]
With respect to the `"0.3"` vs `{ version = "0.3" }`, those two variants are "sufficiently different" that I don't think a take-first thing will be necessary - we can just have a visitor with code to handle a string and map value.

Removing rustc-serialize support makes sense, and will simplify some of the changes as well!

**Comment by alexcrichton** [comment]
Yeah I dunno if removing rustc-serialize is gonna fly, but I think we should at least invert the defaults towards serde at the very least, and perhaps just add back rustc-serialize support in a point release if someone asks for it.

**Comment by dtolnay** [comment]
Rustc-serialize is not going to be making any breaking changes any time soon, so the existing toml 0.2 line will continue to work if someone needs rustc-serialize support.

Dropping rustc-serialize automatically fixes the non-additive cargo feature situation: #128.

> @dtolnay are you thinking like a custom serde attribute on the enum I mentioned to indicate the particular deserialization behavior of "first successful variant wins"? If so, that sounds great!

Yes exactly, as well as an attribute to support Jackson-style internally tagged enums:

```json
{ "type": "X", "other": ..., "fields": ... }
```

**Comment by alexcrichton** [comment]
Ok I've started a branch to upgrade the serde dep: https://github.com/alexcrichton/toml-rs/tree/serde-upgrade.

I'd love to really embrace serde for the 0.3 release, so along those lines what I'd like to do is mainly overhaul the API to match `serde_json` as closely as possible. @dtolnay I'm assuming that `serde_json` represents an idiomatic serde crate in terms of API, right?

Along those lines the changes that need to be done are:

* Implement generic serialization from a type to TOML. Right now you have to encode from a type to a `toml::Value` and then print that `toml::Value` as a string, but that's roundabout and wasteful. I'm about halfway through implementing this on the branch currently.
* Implement generic deserialization from a type to TOML. Right now you also have to deserialize to a `toml::Value` and then deserialize that `toml::Value` into your custom type (again, wasteful). I haven't started this yet but plan to basically mirror the implementation of the JSON parser but for TOML instead.

The main trickiness I'm finding with TOML vs JSON is that with TOML we have to keep state about what table you're in, but I think it's just juggling info as opposed to any major problem. Once those two things are implemented and tests are passing again I'll retrofit unused field checking and such onto the deserializer, I don't think it'll be *too* too hard to do.

**Comment by dtolnay** [comment]
I think the idiomatic pieces are:

- top level modules `::ser` and `::de`
- serializer type `::ser::Serializer`
- serialization functions `::ser::to_*` including `to_writer`
- deserializer type `::de::Deserializer`
- deserialization functions `::de::from_*` including `from_reader`
- none or more of the above re-exported at the top level

The functions are going to vary by format and I don't think we need to aim for perfect consistency across formats as long as the naming makes sense and `to_writer` + `from_reader` are included as a lowest common denominator. Since TOML is UTF-8, `to_string` and `from_str` would probably be convenient.

**Comment by alexcrichton** [comment]
Ok cool, thanks @dtolnay!

**Comment by alexcrichton** [comment]
Ok I think the easy part is done. The `ser` module on that branch should now be mostly complete (modulo bugs). I'll try to start tackling deserialization tonight.

**Comment by alexcrichton** [comment]
@dtolnay oh another clarification: what should deserializers do by default for missing fields? E.g. let's say I'm working with:

```rust
#[derive(Deserialize)]
struct A {
    foo: Option<u32>
}
```

I'd like that an empty TOML document deserializes that successfully (into `foo: None`), but is this something best left to the `#[derive]` implementation or should it be baked in the deserializer itself?

**Comment by dtolnay** [comment]
Deserializers shouldn't care. Treatment of missing fields is up to the Deserialize implementation and `#[derive(Deserialize)]` handles this case for you.

In case you're interested, the relevant code is [here](https://github.com/serde-rs/serde/blob/v0.9.5/serde_derive/src/de.rs#L1019) and [here](https://github.com/serde-rs/serde/blob/v0.9.5/serde/src/de/private.rs#L7).

**Comment by alexcrichton** [comment]
Ok cool, thanks!

**Comment by alexcrichton** [comment]
@dtolnay another question, what's the best way to support custom primitive datatypes? In TOML there's a datetime type, but serde doesn't have inherent methods from that, so what's the best way to decode into a `Value::Datetime`?

**Comment by dtolnay** [comment]
The Serializer and Deserializer are responsible for defining a mapping between TOML and Serde's data model, so what you can do is have them define the mapping something like this:

- TOML string <-> Serde string
- TOML boolean <-> Serde boolean
- etc.
- TOML datetime <-> Serde struct with one field "$datetime" of type Serde string, the string representation of the datetime

Then Serialize and Deserialize are responsible for defining a mapping between Serde's data model and whatever type, so the implementation for toml::Value would do:

- Serde string <-> Value::String
- Serde boolean <-> Value::Boolean
- Serde struct with one field "$datetime" of type Serde string <-> Value::Datetime

And to complete the picture, you should also provide a type toml::Datetime that users can use inside their structs to mean that a datetime is required.

- Serde struct with one field "$datetime" of type Serde string <-> toml::Datetime

**Comment by alexcrichton** [comment]
Ok cool makes sense to me, thanks @dtolnay!

**Comment by alexcrichton** [comment]
@dtolnay hm so then how are these disambiguated?

```toml
foo = 2017-02-04
foo = { '$datetime' = '2017-02-04' }
```

**Comment by dtolnay** [comment]
They wouldn't be, both of those would deserialize as Value::Datetime. Let's fix it by picking a different mapping.

One thing working to our advantage here is that TOML tables have string keys, so we can assign special meaning to Serde maps with non-string keys. For example the Deserializer could treat `2017-02-04` as (in fake JSON syntax) `{["datetime"]: "2017-02-04"}` where the map key is a Serde sequence containing a Serde string. Or keep it simple and use an integer - `{0: "2017-02-04"}` where we declare that 0 stands for datetime.

Basically the [`Visitor`](https://docs.serde.rs/serde/de/trait.Visitor.html) trait is our playground and we just need to define some pattern that means datetime. During deserialization the Deserializer will turn datetimes into that pattern and the Deserialize implementation of Value or toml::Datetime will recognize that pattern and understand it as a datetime. Then during serialization the Serialize implementation of Value or toml::Datetime will recreate the same pattern against the Serializer interface, and the TOML Serializer will recognize the pattern and render it as a datetime.

**Comment by alexcrichton** [comment]
Ok cool that makes sense. Sorry if this is a lot of questions! I've got like a writer's block with implementing this parser...

**Comment by dtolnay** [comment]
I wish I had an example to point you to but I think this is uncharted territory! I am excited to get it working.

**Comment by alexcrichton** [comment]
Ok after a last final push this weekend I think the end is in sight. I'm down to fixing up edge case test failures here and there but I think all necessary infrastructure is in place.

@dtolnay another question. I'd like for [this example](https://github.com/alexcrichton/toml-rs/blob/serde-upgrade/examples/decode.rs) to continue to work, namely that if a field is `Option<T>` and it's not present in the input it decodes as `None`. This seems like something I need to presumably handle in the deserializer because that test case is generating an error right now, but I'm curious where such logic would go?

**Comment by dtolnay** [comment]
Your Deserializer implementations need a deserialize_option method that calls visitor.visit_none() or visitor.visit_some(self) depending on whether the input contains null. [Here is serde_json's.](https://github.com/serde-rs/json/blob/v0.9.5/json/src/de.rs#L586-L599). It looks like TOML doesn't have null so you can unconditionally call visitor.visit_some(self) and missing Option fields will get treated as None by default.

```rust
fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Error>
    where V: de::Visitor
{
    visitor.visit_some(self)
}
```

**Comment by alexcrichton** [comment]
Ok! I think I've got everything working on the [serde-upgrade branch](https://github.com/alexcrichton/toml-rs/tree/serde-upgrade). I believe I've ported over all previous tests and examples and everything's working great!

The last piece missing is the ability to detect unused fields. @dtolnay do you know if there's a good way to handle that? It looks like with the default `Deserialize` implementation structs will slurp up all the fields they're given and just ignore unknown ones (good!), but for Cargo at least we've enjoyed the ability to print out those ignored fields to catch typos and such.

I was thinking originally I could implement `Deserializer for &mut Value` and just pluck out required fields, but I think the current interface means that the value has to hand all of its field to the provided visitor, right?

If this is too difficult to do today with serde I'm also ok going ahead and releasing toml 0.3 without it. We can always upgrade Cargo at a later date! (I don't think anyone else is using this feature)

**Comment by alexcrichton** [comment]
Ok, I've posted a PR: https://github.com/alexcrichton/toml-rs/pull/137

**Comment by dtolnay** [comment]
The `deserialize_ignored_any` method tells us when a Deserialize implementation is ignoring a field. I'll put together a demo of a Deserializer that captures the path of every ignored field. It should be doable generically and efficiently across all data formats.

**Comment by dtolnay** [comment]
I put together some demo code as promised - see #138. Let me know whether this accomplishes everything you would want.

I will start reviewing #137 now.

**Comment by alexcrichton** [comment]
@dtolnay you're the best ❤️! 

I'll polish up #137, land tonight, and start transitioning Cargo pronto

**Comment by dtolnay** [comment]
I added you as a collaborator on the serde-ignored repo and owner on crates.io so go ahead and make any changes you want. I might be busy tonight so I don't want to be a roadblock on that stuff.

**Comment by alexcrichton** [comment]
Fixed in #137

**Comment by alexcrichton** [comment]
Ok, 0.3.0 is now published!
